src/wp-includes/shortcodes.php

```diff
@@ -221,6 +221,12 @@ function do_shortcode( $content, $ignore_html = false ) {
 		return $content;
 	}
 
+	$callback = static function() {
+		return 'do_shortcode';
+	};
+
+	add_filter( 'wp_get_attachment_image_context', $callback );
```

Comment:
Shortcodes are sometimes nested, particularly by plugins that make more complex usage of shortcodes. We should make sure that adding the filter here and removing it below doesn't cause weird behaviors.

I think this code would be more error-proof if we did the following:
* Only add the filter callback if it's not already added to the filter hook.
* Below, only remove the filter callback if it was added in the same `do_shortcode()` execution.
* To ensure that we reference the same callback throughout different `do_shortcode()` calls, we could either make `$callback` a static variable so that it always references the same closure or replace it with a private function (or rather _marked_ as private) like `_return_do_shortcode_context()`.

Reply:
Some unit tests with nest shortcodes would be handy. Also a test for gallery and caption shortcodes, as these are used a lot in wild.